src/Raven.Server/Documents/Document.cs

```diff
@@ -13,6 +14,108 @@ public class TimeSeriesStream
         public IEnumerable<DynamicJsonValue> TimeSeries;
         public string Key;
     }
+
+    public sealed class QueriedDocument : Document, IResetSupport<List<QueriedDocument>>
+    {
+        private static readonly QueriedDocument Empty = new();
+        
+        private int _refCount = 1;
+        
+        
+        private QueriedDocument _parent;
+        private QueriedDocument _child;
+        private List<QueriedDocument> _childs;
+        public bool CanDispose => _refCount <= 1;
+
+        public void AddChild(QueriedDocument child)
+        {
+            if (_child == null)
+            {
+                _child = child;
+            }
+            else
+            {
+                if (ReferenceEquals(Empty, _child) == false)
+                {
+                    _childs ??= new(2);
+                    _childs.Add(_child);
+                    _child = Empty;
+                }
+                
+                _childs.Add(child);
+            }
+            
+            child._refCount++;
+        }
+        
+        public new TDocument Clone<TDocument>(JsonOperationContext context)
+            where TDocument : Document, new()
+        {
+            var data = Data.Clone(context);
+            return CloneWith<TDocument>(context, data);
+        }
+
+        public new TDocument CloneWith<TDocument>(JsonOperationContext context, BlittableJsonReaderObject newData)
+            where TDocument : Document, new()
+
+        {
+            if (typeof(TDocument) != typeof(QueriedDocument))
+                return base.Clone<TDocument>(context);
+            _refCount++;
+            var cloned = base.CloneWith<TDocument>(context, newData);
+            ((QueriedDocument)(object)cloned)._parent = this;
+            AddChild((QueriedDocument)(object)cloned);
+            return cloned;
+        }
+
+        public void RemoveChild(QueriedDocument child)
```

Comment:
See my above comments about the usage of "referenced" / "referencing" instead of "child" / "parent" terminology